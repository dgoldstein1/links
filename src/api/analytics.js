import axios from "axios";

import { makeRequest } from "./utils";
/**
 * gets user's ip address and then posts go website analytics endpoint
 * @return: Promise
 * {
 *   success : bool,
 *   data : [{key, value}],
 *   error : string
 * }
 **/
export function postUserVisit() {
  // first get ip address of user
  let geoUrl = encodeURI(
    `/analytics/api/geoIpServer/v1?apiKey=at_Mb3nWUvk1iAL4W97H5Fs1LxAXjRCn&ipAddress=8.8.8.8`
  );
  // add in href, if exists
  return axios
    .get(geoUrl)
    .then(res => {
      // check for failure
      if (!res.data)
        return Promise.resolve({
          success: false,
          error: "Could not get IP address from 3rd party"
        });
      // now post this data to the backend
      return makeRequest({
        method: "post",
        url: "/analytics/server/visits",
        onErrorPrefix: "could not post json to analytics backend",
        body: _formatDataToAnalyticsBackend(res)
      });
    })
    .catch(e => ({
      success: false,
      error: "could not get user IP address: " + e.message
    }));
}

/**
 * formats data
 *   from: https://geo.ipify.org/docs
 *   into: https://github.com/dgoldstein1/websiteanalytics-backend#visits-1
 **/
export function _formatDataToAnalyticsBackend(ipifyResponse) {
  let d = {
    ip: ipifyResponse.ip,
    city: ipifyResponse.location.city,
    country_code: ipifyResponse.location.country,
    latitude: ipifyResponse.location.lat,
    longitude: ipifyResponse.location.lng,
    region_code: ipifyResponse.location.region,
    time_zone: "ipify" + ipifyResponse.location.timezone,
    zip_code: ipifyResponse.location.postalCode
  };
  // add in referrer code
  d.href = new URLSearchParams(window.location.search).get("href");
  return d;
}
