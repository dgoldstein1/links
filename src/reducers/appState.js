import { SET_LOADING, UPDATE_VIEW, SET_FATAL_ERROR } from "../actions/appState";

const initialState = {
  loading: true, // the app is / isn't loading
  fatalError: undefined, // error which  causes app to not load
  view: "splash" // one of "explore", "path", "splash", "about", "settings"
};

const visitReducer = (state = initialState, action) => {
  switch (action.type) {
    case SET_FATAL_ERROR:
      return Object.assign({}, state, {
        fatalError: action.fatalError
      });
    // set when the app is loading
    case SET_LOADING:
      return Object.assign({}, state, {
        loading: action.loading === undefined ? false : action.loading
      });
    // update where the user is currently looking
    case UPDATE_VIEW:
      return Object.assign({}, state, {
        view: action.view
      });
    default:
      return state;
  }
};

export default visitReducer;
