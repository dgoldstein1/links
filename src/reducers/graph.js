import _ from "lodash";

const initialState = {
  rootNode: {}, // root node of graph (what is in search or 'start from')
  selectedNode: {
    // node currently in view
    loading: true,
    error: "",
    description: "",
    img: "",
    node: {}
  },
  targetNode: {}, // target node in path ('ending at..')
  graph: {
    nodes: [],
    edges: []
  },
  maxShortestPaths: 5,
  directedShortestPath: true,
  loading: false,
  error: undefined,
  layout: "hierarchy",
  maxNeighbors: 15,
  topInfo: {
    betweenessEdges: [],
    betweenessNodes: [],
    pageRank: []
  }
};

const graphReducer = (state = initialState, action) => {
  switch (action.type) {
    case "CLEAR_GRAPH":
      return Object.assign({}, state, {
        graph: {
          nodes: [],
          edges: []
        }
      });
    case "SET_DIRECTED_SHORTEST_PATH":
      return Object.assign({}, state, {
        directedShortestPath: action.directedShortestPath
      });
    case "SET_MAX_NEIGHBORS":
      return Object.assign({}, state, {
        maxNeighbors: action.maxNeighbors
      });
    case "SET_GRAPH_ERROR":
      return Object.assign({}, state, {
        error: action.error
      });
    case "SET_GRAPH_LAYOUT":
      return Object.assign({}, state, {
        layout: action.layout
      });
    case "SET_GRAPH_LOADING":
      return Object.assign({}, state, {
        loading: action.loading
      });
    case "SET_ROOT_NODE":
      return Object.assign({}, state, {
        rootNode: _generateRoot(action.node.label, action.node.id)
      });
    case "SET_MAX_SHORTEST_PATHS":
      return Object.assign({}, state, {
        maxShortestPaths: action.maxShortestPaths
      });
    case "SET_ALL_PATHS_UNIQUE":
      return Object.assign({}, state, {
        pathsAreUnique: action.pathsAreUnique
      });

    case "SET_SELECTED_NODE":
      return Object.assign({}, state, {
        selectedNode: {
          ...state.selectedNode,
          node: action.node
        }
      });
    case "SET_SELECTED_NODE_INFO":
      return Object.assign({}, state, {
        selectedNode: {
          ...state.selectedNode,
          description: action.description,
          img: action.img,
          loading: action.loading,
          error: action.error,
          centrality: action.centrality
        }
      });
    case "SET_TARGET_NODE":
      return Object.assign({}, state, {
        targetNode: action.node
      });

    case "SET_GRAPH_PATH":
      // create array of all nodes from entries
      let nodes = action.entries.map(e => ({
        id: e.value,
        label: e.key,
        size: 1,
        color: "red"
      }));
      // create edges from each path in graph
      let edges = [];
      let entriesReverseMap = {};
      action.entries.forEach(e => (entriesReverseMap[e.value] = e.key));
      action.paths.forEach(p => {
        p.forEach((id, i) => {
          // dont add edge if we've reached the end
          if (i === p.length - 1) return;
          // lookup this entry and next entry
          edges.push({
            id: `${id}->-->--${p[i + 1]}`,
            source: id,
            target: p[i + 1],
            color: "red"
          });
        });
      });
      return Object.assign({}, state, {
        graph: { nodes, edges }
      });
    case "ADD_NEIGHBORS_TO_GRAPH":
      // set initial position to source
      action.neighbors = action.neighbors.map(n => ({
        ...n,
        size: 1
      }));
      // get list of edges to add
      let edgesToAdd = [];
      action.neighbors.forEach(n => {
        if (!n.id) console.log(n);
        edgesToAdd.push({
          id: `${action.node.id}->-->--${n.id}`,
          source: action.node.id,
          target: n.id,
          color: "black"
        });
      });
      // add in node to make sure it's in list of nodes
      action.neighbors.push(action.node);
      // return state
      return Object.assign({}, state, {
        graph: {
          nodes: _.unionBy(state.graph.nodes, action.neighbors, "id"),
          edges: _.unionBy(state.graph.edges, edgesToAdd, "id")
        }
      });
    default:
      return state;
  }
};

export function _generateRoot(label, id) {
  return {
    label,
    id,
    x: 0,
    y: 0,
    size: 1
  };
}

export default graphReducer;
