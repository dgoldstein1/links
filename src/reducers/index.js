// redux utils
import { combineReducers, createStore, applyMiddleware } from "redux";
import { createLogger } from "redux-logger";

// reducers in this directory
import appState from "./appState";
import graph from "./graph";

const Reducer = combineReducers({
  appState,
  graph,
});

const logger = createLogger({
  collapsed: true,
  diff: true,
});

let middleware = {};
if (process.env.REACT_APP_ENV === "development") {
  middleware = applyMiddleware(logger);
}

const store = createStore(Reducer, middleware);

export { store };
