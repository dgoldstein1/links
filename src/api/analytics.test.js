import moxios from "moxios";
import { store } from "../reducers";
import sinon from "sinon";
import axios from "axios";
import { _getIpAddress, postUserVisit } from "./analytics";

describe("analytics", () => {
  describe("_getIpAddress", () => {
    beforeEach(() => {
      moxios.install();
    });

    afterEach(() => {
      moxios.uninstall();
    });
    let testTable = [
      {
        name: "normal IP",
        url: `/analytics/api/geoIpServer/check?access_key=7eca814a6de384aab338e110c57fef37`,
        responseCode: 200,
        data: {
          ip: "14.35.263.24"
        }
      },
      {
        name: "deliminted IP",
        url: `/analytics/api/geoIpServer/check?access_key=7eca814a6de384aab338e110c57fef37`,
        responseCode: 200,
        data: {
          ip: "14.35.263.24, 25.362.53.32"
        }
      }
    ];
    testTable.forEach(t => {
      it(t.name, done => {
        let onFulfilled = sinon.spy();
        moxios.stubRequest(t.url, {
          status: t.responseCode,
          response: t.data
        });
        _getIpAddress().then(ip => {
          expect(ip).toMatchSnapshot();
          done();
        });
      });
    });
  });
  describe("postUserVisits", () => {
    beforeEach(() => {
      moxios.install();
    });

    afterEach(() => {
      moxios.uninstall();
    });
    let ipAddress = "14.35.263.24";
    let testTable = [
      {
        name: "normal IP",
        ipUrl: `/analytics/api/geoIpServer/check?access_key=7eca814a6de384aab338e110c57fef37`,
        ipResponse: { ip: ipAddress },
        geoIpUrl: `/analytics/api/geoIpServer/${ipAddress}?access_key=7eca814a6de384aab338e110c57fef37`,
        analyticsServerUrl: "/analytics/server/visits",
        analyticsServerResponse: {
          ip: "205.153.92.177",
          city: "Sterling",
          country_code: "US",
          country_name: "United States",
          latitude: 38.952701568603516,
          longitude: -77.44322967529297,
          metro_code: 0,
          region_code: "VA",
          time_zone: "",
          zip_code: "20166",
          visit_date: "2019-11-27T01:01:23.504356771Z"
        },
        responseCode: 200,
        geoIpResponse: {
          ip: "205.153.92.177",
          type: "ipv4",
          continent_code: "NA",
          continent_name: "North America",
          country_code: "US",
          country_name: "United States",
          region_code: "VA",
          region_name: "Virginia",
          city: "Sterling",
          zip: "20166",
          latitude: 38.952701568603516,
          longitude: -77.44322967529297,
          location: {
            geoname_id: 4787534,
            capital: "Washington D.C.",
            languages: [{ code: "en", name: "English", native: "English" }],
            country_flag: "http://assets.ipstack.com/flags/us.svg",
            country_flag_emoji: "\ud83c\uddfa\ud83c\uddf8",
            country_flag_emoji_unicode: "U+1F1FA U+1F1F8",
            calling_code: "1",
            is_eu: false
          }
        }
      },
      {
        name: "bad geoIpRequest",
        ipUrl: `/analytics/api/geoIpServer/check?access_key=7eca814a6de384aab338e110c57fef37`,
        ipResponse: { ip: ipAddress },
        geoIpUrl: `/analytics/api/geoIpServer/${ipAddress}?access_key=7eca814a6de384aab338e110c57fef37`,
        analyticsServerUrl: "/analytics/server/visits",
        analyticsServerResponse: {
          ip: "205.153.92.177",
          city: "Sterling",
          country_code: "US",
          country_name: "United States",
          latitude: 38.952701568603516,
          longitude: -77.44322967529297,
          metro_code: 0,
          region_code: "VA",
          time_zone: "",
          zip_code: "20166",
          visit_date: "2019-11-27T01:01:23.504356771Z"
        },
        responseCode: 500,
        geoIpResponse: {}
      }
    ];
    testTable.forEach(t => {
      it(t.name, done => {
        let onFulfilled = sinon.spy();
        // ip address
        moxios.stubRequest(t.ipUrl, {
          status: 200,
          response: t.ipResponse
        });
        // geoIpCheck
        moxios.stubRequest(t.geoIpUrl, {
          status: t.responseCode,
          response: t.geoIpResponse
        });
        // analytics server "/visits" post
        moxios.stubRequest(t.analyticsServerUrl, {
          status: 200,
          response: t.analyticsServerResponse
        });
        postUserVisit().then(r => {
          expect(r).toMatchSnapshot();
          done();
        });
      });
    });
  });
});
