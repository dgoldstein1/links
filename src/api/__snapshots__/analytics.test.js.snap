// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`analytics _formatDataToAnalyticsBackend normal response 1`] = `
Object {
  "city": "Hastings",
  "country_code": "US",
  "href": null,
  "ip": "96.72.50.249",
  "latitude": 44.7443,
  "longitude": -92.8514,
  "region_code": "Minnesota",
  "time_zone": "ipify-06:00",
  "zip_code": "55033",
}
`;

exports[`analytics postUserVisits bad geoIpRequest 1`] = `
Object {
  "error": "could not get user IP address: Request failed with status code 500",
  "success": false,
}
`;

exports[`analytics postUserVisits normal IP 1`] = `
Object {
  "data": Object {
    "city": "Sterling",
    "country_code": "US",
    "country_name": "United States",
    "ip": "205.153.92.177",
    "latitude": 38.952701568603516,
    "longitude": -77.44322967529297,
    "metro_code": 0,
    "region_code": "VA",
    "time_zone": "",
    "visit_date": "2019-11-27T01:01:23.504356771Z",
    "zip_code": "20166",
  },
  "success": true,
}
`;
